### ISSUE TITLE
[YSQL][randgen] Wrong results from correlated scalar subquery using HashAggregate plan (a vanilla Postgres issue)

### ISSUE THREAD

**Issue by mtakahar** [yugabyte/yugabyte-db#21012]
Jira Link: [DB-9986](https://yugabyte.atlassian.net/browse/DB-9986)
### Description

Problem
---
**Update: It turns out this is a bug in postgres that does not manifest itself on YB. ([details](https://github.com/yugabyte/yugabyte-db/issues/21012#issuecomment-1942103526))**
* ~Correlated scalar subquery with aggregate in the SELECT-list incorrectly returns non-null values.~
* ~No BNL involved.~
* ~Reproduces on recent master (0214819b32).~

Test Case
---
* Further simplified test case that I put in the report for the postgres devs:
```
postgres=# select version();
                                                           version          
                                                 
------------------------------------------------------------------------------------------------------------------------------
 PostgreSQL 15.7 (Homebrew) on x86_64-apple-darwin22.6.0, compiled by Apple
clang version 15.0.0 (clang-1500.1.0.2.5), 64-bit
(1 row)



postgres=# CREATE TABLE b (col_int int);
CREATE TABLE cc (col_int int);



INSERT INTO b values (1);
INSERT INTO cc values (null), (1), (2);



CREATE TABLE
postgres=# CREATE TABLE
postgres=# postgres=# INSERT 0 1
postgres=# INSERT 0 3
postgres=# postgres=# \pset null '<null>'
Null display is "<null>".
postgres=# SET enable_hashagg = ON; SET enable_sort = OFF;
SET
SET
postgres=# SELECT (  SELECT  DISTINCT  MIN( col_int  )  FROM B  AS
SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_int  = table1 .col_int  ), table1
.col_int  FROM CC  AS table1    ;
  min   | col_int 
--------+---------
 <null> |  <null>
 <null> |       1        <------- *** wrong ***
 <null> |       2
(3 rows)



postgres=# EXPLAIN SELECT (  SELECT  DISTINCT  MIN( col_int  )  FROM B  AS
SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_int  = table1 .col_int  ), table1
.col_int  FROM CC  AS table1    ;
                                       QUERY PLAN                           
           
----------------------------------------------------------------------------------------
 Seq Scan on cc table1  (cost=0.00..8306.82 rows=2550 width=8)
   SubPlan 2
     ->  HashAggregate  (cost=3.23..3.24 rows=1 width=4)
           Group Key: $1
           InitPlan 1 (returns $1)
             ->  Limit  (cost=0.00..3.22 rows=1 width=4)
                   ->  Seq Scan on b subquery1_t1  (cost=0.00..41.88 rows=13
width=4)
                         Filter: ((col_int IS NOT NULL) AND (col_int =
table1.col_int))
           ->  Result  (cost=3.22..3.23 rows=1 width=4)
(9 rows)



postgres=# SET enable_hashagg = OFF; SET enable_sort = ON;
SET
SET
postgres=# SELECT (  SELECT  DISTINCT  MIN( col_int  )  FROM B  AS
SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_int  = table1 .col_int  ), table1
.col_int  FROM CC  AS table1    ;
  min   | col_int 
--------+---------
 <null> |  <null>
      1 |       1        <------- *** correct ***
 <null> |       2
(3 rows)



postgres=# EXPLAIN SELECT (  SELECT  DISTINCT  MIN( col_int  )  FROM B  AS
SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_int  = table1 .col_int  ), table1
.col_int  FROM CC  AS table1    ;
                                       QUERY PLAN                           
           
----------------------------------------------------------------------------------------
 Seq Scan on cc table1  (cost=0.00..8319.57 rows=2550 width=8)
   SubPlan 2
     ->  Unique  (cost=3.24..3.25 rows=1 width=4)
           InitPlan 1 (returns $1)
             ->  Limit  (cost=0.00..3.22 rows=1 width=4)
                   ->  Seq Scan on b subquery1_t1  (cost=0.00..41.88 rows=13
width=4)
                         Filter: ((col_int IS NOT NULL) AND (col_int =
table1.col_int))
           ->  Sort  (cost=3.24..3.25 rows=1 width=4)
                 Sort Key: ($1)
                 ->  Result  (cost=3.22..3.23 rows=1 width=4)
(10 rows)
```

### Issue Type

kind/bug

### Warning: Please confirm that this issue does not contain any sensitive information

- [X] I confirm this issue does not contain any sensitive information.

[DB-9986]: https://yugabyte.atlassian.net/browse/DB-9986?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by tanujnay112** [comment]
Seems to happen on non-colocated db's as well.

However, it seems like Postgres15 itself can't make up its mind about the correct result.
```
postgres=# SELECT ( SELECT DISTINCT MAX( SUBQUERY1_t1.pk ) AS SUBQUERY1_field1 FROM B AS SUBQUERY1_t1 WHERE SUBQUERY1_t1.col_varchar_key != table1.col_varchar_key ) AS field1, table1.col_varchar_key AS field2 FROM ( CC AS table1 LEFT OUTER JOIN ( ( CC AS table2 RIGHT JOIN CC AS table3 ON (table3.col_varchar_key = table2.col_varchar_nokey ) ) ) ON (table3.col_varchar_key = table2.col_varchar_key ) ) WHERE ( EXISTS ( SELECT SUBQUERY2_t1.col_int_key AS SUBQUERY2_field1 FROM ( B AS SUBQUERY2_t1 LEFT JOIN ( C AS SUBQUERY2_t2 LEFT OUTER JOIN C AS SUBQUERY2_t3 ON (SUBQUERY2_t3.col_varchar_key = SUBQUERY2_t2.col_varchar_key ) ) ON (SUBQUERY2_t3.col_int_key = SUBQUERY2_t2.pk ) ) WHERE SUBQUERY2_t2.col_int_nokey <= table2.col_int_nokey ) ) OR ( table1.pk > 243 AND table1.pk < ( 6 + 23 ) OR table1.pk IN (178) ) OR table1.pk IS NULL ORDER BY field2  NULLS FIRST , field2  NULLS FIRST , field1  NULLS FIRST , field2  NULLS FIRST  ;

.
.
.
.
.
        | y
        | y
        | y
        | y
        | y
        | y
        | y
(460 rows)
<Expected result in GHI>

postgres=# set enable_indexscan to false;
.
.
.
.
.
        | x
        | x
        | x
        | x
        | x
        | x
        | x
        | x
        | x
        | x
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
      1 | y
(460 rows)
```

**Comment by mtakahar** [comment]
Simplified Test Case
---
I was able to get [the RQG query simplifier](https://github.com/yugabyte/randgen/blob/master/util/simplify-query.pl) working on YB & PG and got this:
```
SELECT (  SELECT  DISTINCT  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;
```

PG15 returns different results from different plans, the behavior @tanujnay112 discovered with the original test case:

With IndexScan
---
```
PG15 test# \pset null '<null>'
Null display is "<null>".
PG15 test# SELECT (  SELECT  DISTINCT  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

  max   
--------
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
 <null>
(20 rows)

PG15 test# EXPLAIN SELECT (  SELECT  DISTINCT  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

                                                  QUERY PLAN                                                   
---------------------------------------------------------------------------------------------------------------
 Seq Scan on cc table1  (cost=0.00..230.66 rows=850 width=4)
   SubPlan 2
     ->  HashAggregate  (cost=0.24..0.25 rows=1 width=4)
           Group Key: $1
           InitPlan 1 (returns $1)
             ->  Limit  (cost=0.15..0.23 rows=1 width=4)
                   ->  Index Scan Backward using b_pkey on b subquery1_t1  (cost=0.15..65.07 rows=842 width=4)
                         Index Cond: (pk IS NOT NULL)
                         Filter: ((col_varchar_key)::text <> (table1.col_varchar_key)::text)
           ->  Result  (cost=0.23..0.24 rows=1 width=4)
(10 rows)
```
With SeqScan
---
```
PG15 test# /*+ Set(enable_indexscan off) */ SELECT (  SELECT  DISTINCT  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

  max   
--------
 <null>
      1
      1
      1
      1
      1
      1
 <null>
      1
      1
      1
      1
 <null>
      1
      1
 <null>
      1
      1
      1
      1
(20 rows)

PG15 test# EXPLAIN /*+ Set(enable_indexscan off) */ SELECT (  SELECT  DISTINCT  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

                                           QUERY PLAN                                            
-------------------------------------------------------------------------------------------------
 Seq Scan on cc table1  (cost=0.00..19370.88 rows=850 width=4)
   SubPlan 1
     ->  Unique  (cost=22.76..22.77 rows=1 width=4)
           ->  Sort  (cost=22.76..22.77 rows=1 width=4)
                 Sort Key: (max(subquery1_t1.pk))
                 ->  Aggregate  (cost=22.74..22.75 rows=1 width=4)
                       ->  Seq Scan on b subquery1_t1  (cost=0.00..20.62 rows=846 width=4)
                             Filter: ((col_varchar_key)::text <> (table1.col_varchar_key)::text)
(8 rows)
```

**Comment by mtakahar** [comment]
PG15 returns the same results without `DISTINCT` although it is just redundant because the subquery has an aggregate function without GROUP BY clause, which always return exactly 1 row.

With IndexScan
---
```
PG15 test# /*+ Set(enable_indexscan on) */ SELECT (  SELECT  /*DISTINCT*/  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

  max   
--------
 <null>
      1
      1
      1
      1
      1
      1
 <null>
      1
      1
      1
      1
 <null>
      1
      1
 <null>
      1
      1
      1
      1
(20 rows)

PG15 test# EXPLAIN /*+ Set(enable_indexscan on) */ SELECT (  SELECT  /*DISTINCT*/  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

                                                  QUERY PLAN                                                   
---------------------------------------------------------------------------------------------------------------
 Seq Scan on cc table1  (cost=0.00..220.04 rows=850 width=4)
   SubPlan 2
     ->  Result  (cost=0.23..0.24 rows=1 width=4)
           InitPlan 1 (returns $1)
             ->  Limit  (cost=0.15..0.23 rows=1 width=4)
                   ->  Index Scan Backward using b_pkey on b subquery1_t1  (cost=0.15..65.07 rows=842 width=4)
                         Index Cond: (pk IS NOT NULL)
                         Filter: ((col_varchar_key)::text <> (table1.col_varchar_key)::text)
(8 rows)
```
With SeqScan
---
```
PG15 test# /*+ Set(enable_indexscan off) */ SELECT (  SELECT  /*DISTINCT*/  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

  max   
--------
 <null>
      1
      1
      1
      1
      1
      1
 <null>
      1
      1
      1
      1
 <null>
      1
      1
 <null>
      1
      1
      1
      1
(20 rows)

PG15 test# EXPLAIN /*+ Set(enable_indexscan off) */ SELECT (  SELECT  /*DISTINCT*/  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

                                     QUERY PLAN                                      
-------------------------------------------------------------------------------------
 Seq Scan on cc table1  (cost=0.00..19356.00 rows=850 width=4)
   SubPlan 1
     ->  Aggregate  (cost=22.74..22.75 rows=1 width=4)
           ->  Seq Scan on b subquery1_t1  (cost=0.00..20.62 rows=846 width=4)
                 Filter: ((col_varchar_key)::text <> (table1.col_varchar_key)::text)
(5 rows)
```

**Comment by mtakahar** [comment]
Expected Behavior
---
The following query returns each value of the subquery, the join condition and each individual columns in the join condition. The results from the simplified query modified to show table1.col_varchar_key are consistent, so the results from YB and PG15 w/SeqScan are correct.
```
SELECT
  B_pk_max AS "MAX(B.pk)",
  sq.col_varchar_key != table1.col_varchar_key
      AS "sq.col_varchar_key!=table1.col_varchar_key",
  table1.col_varchar_key AS "table1.col_varchar_key",
  sq.col_varchar_key AS "B.col_varchar_key"
FROM CC AS table1
     LEFT JOIN (SELECT MAX(B.pk) AS b_pk_max, B.col_varchar_key FROM B GROUP BY B.col_varchar_key) AS sq
       ON sq.col_varchar_key != table1.col_varchar_key
ORDER BY table1.col_varchar_key;

 MAX(B.pk) | sq.col_varchar_key!=table1.col_varchar_key | table1.col_varchar_key | B.col_varchar_key 
-----------+--------------------------------------------+------------------------+-------------------
         1 | t                                          | c                      | x
         1 | t                                          | d                      | x
         1 | t                                          | d                      | x
         1 | t                                          | f                      | x
         1 | t                                          | f                      | x
         1 | t                                          | g                      | x
         1 | t                                          | j                      | x
         1 | t                                          | m                      | x
         1 | t                                          | m                      | x
         1 | t                                          | o                      | x
         1 | t                                          | p                      | x
         1 | t                                          | q                      | x
         1 | t                                          | r                      | x
         1 | t                                          | u                      | x
         1 | t                                          | w                      | x
    <null> | <null>                                     | x                      | <null>
    <null> | <null>                                     | x                      | <null>
         1 | t                                          | y                      | x
    <null> | <null>                                     | <null>                 | <null>
    <null> | <null>                                     | <null>                 | <null>
(20 rows)
```
Simplified Query Modified to Show table1.col_varchar_key:
```
SELECT (  SELECT  DISTINCT  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  ), table1.col_varchar_key  FROM CC  AS table1   ORDER BY table1.col_varchar_key;

  max   | col_varchar_key 
--------+-----------------
      1 | c
      1 | d
      1 | d
      1 | f
      1 | f
      1 | g
      1 | j
      1 | m
      1 | m
      1 | o
      1 | p
      1 | q
      1 | r
      1 | u
      1 | w
 <null> | x
 <null> | x
      1 | y
 <null> | <null>
 <null> | <null>
(20 rows)
```

**Comment by mtakahar** [comment]
YB returns correct results with or without IndexScan:
```
test=# /*+ Set(enable_indexscan on) */ SELECT (  SELECT  DISTINCT  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

  max   
--------
 <null>
      1
      1
      1
      1
      1
      1
 <null>
      1
      1
      1
      1
 <null>
      1
      1
 <null>
      1
      1
      1
      1
(20 rows)

test=# EXPLAIN (COSTS OFF)
/*+ Set(enable_indexscan on) */ SELECT (  SELECT  DISTINCT  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

                                             QUERY PLAN                                             
----------------------------------------------------------------------------------------------------
 Seq Scan on cc table1
   SubPlan 2
     ->  Unique
           InitPlan 1 (returns $1)
             ->  Limit
                   ->  Index Scan Backward using b_pkey on b subquery1_t1
                         Index Cond: (pk IS NOT NULL)
                         Remote Filter: ((col_varchar_key)::text <> (table1.col_varchar_key)::text)
           ->  Sort
                 Sort Key: ($1)
                 ->  Result
(11 rows)

test=# /*+ Set(enable_indexscan off) */ SELECT (  SELECT  DISTINCT  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

  max   
--------
 <null>
      1
      1
      1
      1
      1
      1
 <null>
      1
      1
      1
      1
 <null>
      1
      1
 <null>
      1
      1
      1
      1
(20 rows)

test=# EXPLAIN (COSTS OFF)
/*+ Set(enable_indexscan off) */ SELECT (  SELECT  DISTINCT  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

                                               QUERY PLAN                                               
--------------------------------------------------------------------------------------------------------
 Seq Scan on cc table1
   SubPlan 1
     ->  Unique
           ->  Sort
                 Sort Key: (max(subquery1_t1.pk))
                 ->  Finalize Aggregate
                       ->  Seq Scan on b subquery1_t1
                             Remote Filter: ((col_varchar_key)::text <> (table1.col_varchar_key)::text)
                             Partial Aggregate: true
(9 rows)
```

**Comment by mtakahar** [comment]
This is a Postgres issue but not reproducible on YB.

**Comment by mtakahar** [comment]
Found a GUC/hint combination for reproducing it on YB.
```
CREATE TABLE public.b (
    pk integer NOT NULL,
    col_varchar_key character varying(1),
    CONSTRAINT b_pkey PRIMARY KEY(pk ASC)
);

CREATE TABLE public.cc (
    col_varchar_key character varying(1)
);

COPY public.b (pk, col_varchar_key) FROM stdin;
1	x
\.

COPY public.cc (col_varchar_key) FROM stdin;
d
f
c
m
o
\N
g
d
r
x
f
j
m
x
u
w
q
p
y
\N
\.
```

```
psql (15.1, server 11.2-YB-2.21.1.0-b0)
Type "help" for help.

YB test# /*+ Set(enable_sort OFF) IndexScan(subquery1_t1 b_pkey) */SELECT (  SELECT  DISTINCT  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

 max 
-----
   1
   1
   1
   1
   1
   1
   1
   1
   1
   1
   1
   1
   1
   1
   1
   1
   1
   1
   1
   1
(20 rows)

YB test# EXPLAIN (COSTS OFF)
/*+ Set(enable_sort OFF) IndexScan(subquery1_t1 b_pkey) */SELECT (  SELECT  DISTINCT  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

                                             QUERY PLAN                                             
----------------------------------------------------------------------------------------------------
 Seq Scan on cc table1
   SubPlan 2
     ->  HashAggregate
           Group Key: $1
           InitPlan 1 (returns $1)
             ->  Limit
                   ->  Index Scan Backward using b_pkey on b subquery1_t1
                         Index Cond: (pk IS NOT NULL)
                         Remote Filter: ((col_varchar_key)::text <> (table1.col_varchar_key)::text)
           ->  Result
(10 rows)

YB test# /*+ Set(enable_hashagg OFF) IndexScan(subquery1_t1 b_pkey) */SELECT (  SELECT  DISTINCT  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

  max   
--------
      1
      1
 <null>
      1
      1
      1
      1
      1
      1
      1
 <null>
 <null>
      1
      1
      1
      1
      1
 <null>
      1
      1
(20 rows)

YB test# EXPLAIN (COSTS OFF)
/*+ Set(enable_hashagg OFF) IndexScan(subquery1_t1 b_pkey) */SELECT (  SELECT  DISTINCT  MAX( pk  )  FROM B  AS SUBQUERY1_t1  WHERE SUBQUERY1_t1 .col_varchar_key  != table1 .col_varchar_key  )  FROM CC  AS table1    ;

                                             QUERY PLAN                                             
----------------------------------------------------------------------------------------------------
 Seq Scan on cc table1
   SubPlan 2
     ->  Unique
           InitPlan 1 (returns $1)
             ->  Limit
                   ->  Index Scan Backward using b_pkey on b subquery1_t1
                         Index Cond: (pk IS NOT NULL)
                         Remote Filter: ((col_varchar_key)::text <> (table1.col_varchar_key)::text)
           ->  Sort
                 Sort Key: ($1)
                 ->  Result
(11 rows)
```

**Comment by mtakahar** [comment]
Reported this issue via PG bug reporting form:
BUG 18465: Wrong results from SELECT DISTINCT MIN in scalar subquery using HashAggregate

**Comment by mtakahar** [comment]
Got a response:

> From: | Dmitry Dolgov <9erthalion6(at)gmail(dot)com>
>...
> Thanks, I could reproduce this in v15 and v16. Interestingly enough it
> was fixed in d0d44049d1, but looks like due to lack of error reports the
> fix was applied only on the main branch. I assume this one makes it
> necessary to back-patch the fix.
